forem/forem
# Reusable user admin modals

## What type of PR is this? (check all applicable)


## Description

This PR generalises the way we use modals for forms in the admin member area so that we can use them also on the index page, and (in the future) for bulk actions. To summarise the approach taken:

- Pages which use these modals can now invoke `showUserModal` from `editUserModals.js` on a click event. This will check for a matching modal trigger (via the `dataset` we were already using) and present/initialize the content if a matching modal exists
- Content inside the shared modals isn't fully initialized before this JS runs. The JS populates any form actions, and user-specific data like names, credit balance, and so on. All required data is expected in the trigger buttons `dataset`, so we are still able to take advantage of Rails helpers/data to populate the views. 
- Some modals from the member detail view are not currently planned to be used elsewhere, so I haven't generalised them/added initializer functions for them (although their presentation is still triggered by the same helper function)

This approach allows us to have _one_ hidden chunk of modal HTML on the page, regardless of how many users we want to use it with (avoiding having to create 50 hidden chunks of modal HTML on the member index page). It also sets us up for making sure a bulk action can be attached to the given form too.

## Related Tickets & Documents



- Closes https://github.com/forem/forem/issues/17479

## QA Instructions, Screenshots, Recordings

As this is a refactor, we already had a lot of Cypress tests in place to check this functionality 🎉  But to check manually:

- Log in as an admin and visit the member manager section
- Choose a user from the list
- Experiment with the functions on the page - adding roles, credits, organizations, etc, and check all forms continue to work as expected

### UI accessibility concerns?

I don't think so. We're still using our Preact modal with baked in keyboard access etc

## Added/updated tests?

      have not been included_

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
- [X] This change does not need to be communicated, and this is why not: no change to user experience



## [optional] What gif best describes this PR or how it makes you feel?

![it's different](https://media.giphy.com/media/l378xw31ztiHQi1YA/giphy.gif)
